core/windows_and_tabs/windows_and_tabs.py

```diff
@@ -0,0 +1,48 @@
+from talon import Context, actions, ui
+
+ctx = Context()
+
+
+@ctx.action_class("app")
+class AppActions:
+    def window_previous():
+        cycle_windows(ui.active_app(), -1)
+
+    def window_next():
+        cycle_windows(ui.active_app(), 1)
+
+
+def cycle_windows(app: ui.App, diff: int):
+    """Cycle windows backwards or forwards for the given application"""
+    active = app.active_window
+    windows = [w for w in app.windows() if w == active or is_window_valid(w)]
+    windows.sort(key=lambda w: w.id)
+    current = windows.index(active)
+    max = len(windows) - 1
+    i = cycle(current + diff, 0, max)
+
+    while i != current:
+        try:
+            actions.user.switcher_focus_window(windows[i])
+            break
+        except Exception:
+            i = cycle(i + diff, 0, max)
+
+
+def is_window_valid(window: ui.Window) -> bool:
+    """Returns true if this window is valid for focusing"""
+    return (
+        not window.hidden
+        and window.title != ""
```

Comment:
```suggestion
        # On Windows, there are many fake windows with empty titles -- this excludes them.
        and window.title != ""
```

Reply:
We _might_ just want to scope this check to only windows, since that's the only platform where we've seen this issue. I guess it's a mix of "specific workaround for a known problem that plagues windows" versus "general useful heuristic across platforms". I'm not sure if it would be normal for a macOS window to not have a title. @nriley thoughts?